The patch was taken from solaris-userland.

The logic here is that we want to (a) pass correct args to gcc [which would die
horribly, otherwise], and (b) set up the correct RPATH so that we don't try to
link 32- and 64-bit objects together.

Expected to be contributed upstream at some point.


--- cffi-1.14.0/testing/embedding/test_basic.py
+++ cffi-1.14.0/testing/embedding/test_basic.py
@@ -1,3 +1,4 @@
+import platform
 import py
 import sys, os, re
 import shutil, subprocess, time
@@ -7,6 +8,7 @@ import cffi
 
 local_dir = os.path.dirname(os.path.abspath(__file__))
 _link_error = '?'
+bits = platform.architecture()[0][0:2]
 
 def check_lib_python_found(tmpdir):
     global _link_error
@@ -129,11 +131,19 @@ if sys.platform == 'win32':
                 extra_preargs.append('/MANIFEST')
                 debug = False    # you need to install extra stuff
                                  # for this to work
+            elif sys.platform == 'sunos5':
+                extra_preargs.append("-m%s" % bits)
             elif threads:
                 extra_preargs.append('-pthread')
             objects = c.compile([filename], macros=sorted(defines.items()),
-                                debug=debug)
-            c.link_executable(objects + modules, name, extra_preargs=extra_preargs)
+                                extra_preargs=extra_preargs, debug=debug)
+
+            if sys.platform == 'sunos5':
+                library_dirs = os.path.join("/usr/lib", bits)
+                c.link_executable(objects + modules, name, library_dirs=[library_dirs],
+                                  extra_preargs=extra_preargs)
+            else:
+                c.link_executable(objects + modules, name, extra_preargs=extra_preargs)
         finally:
             os.chdir(curdir)
 

